Advance on minecraft file handling library

developpeur2000 opened the pull request:
Hi,

I've been working on a minecraft map generating engine for the past months and chose your library as a starting base to manage the output minecraft data.
I've finished some functionality like saving the files, adding entities (not all though), etc.
I've corrected a few bugs as I've been testing extensively all the loading and saving
Going through all the changes I've done might take a while, know that I tried to fit to the structure you had set, and keep the code in the same style as you did.
I have only used the base part of your code, and didn't use the tools part.
I will still work on my engine for the next month, so I'll probably make a few other contributions.
I hope I will have time to make some proper javadoc commenting so we can generate a proper documentation.

cheers !
david

developpeur2000 committed:
Added method to unload regions to free some memory
Added Biome definitions
Implemented creation of empty chunks with default data that are save-ready
Implemented saving of all blocks
Added Block Data, some Block Data constants are still to be entered
Added a few assertions
Added loading and saving of level.dat file
Correction of a few bugs, mainly related to loading (bad binary data reading)
Added support for 7zip and gzip compression with choice capability

Commit:
added a readonly option to World

Commit:
Set public the number of blocks at each level
Added methods to create and retrieve entities
Implemented classes corresponding to items and a few mobs (villager has to be completed)
Saving entities has yet to be tested

Commit:
Implemented saving of annotated classes (list and compounds)
Adjusted some initializing that was missing in Entity, Item and Mob

Commit:
bug correction that would store Y of chunks as integer instead of byte - minecraft didn't seem to care much though

Commit:
- Add the possibility to save empty NBT list tags as a specific type, as some are saved by minecraft as empty byte list instead of null type list
- Define blocks skylight (in actual blocks and in part of chunks without defined blocks) as the maximum value so third party software as overviewer can render the blocks properly
- Corrected the size indicated for chunk compressed data as an ending null byte is needed for some third party software
- Corrected the reading of half byte stocked section data (oven and odd values were inverted)

Commit:
setting the lightcalculated flag to true to avoid long opening times of generated maps
todo : actually do the calculation instead of putting full skylight and no block light everywhere

Commit:
transferred the management of blocklight and skylight into sections and chunks instead of blocks, as this is a global computing from chunks (at least)
implemented a very simple skylight computing so the surface will look ok as long as there is no transparent blocks on the surface

Commit:
- Added the definition of a bitwise filter for tree types of saplings
- added the support of saving a region that already has a file, by saving to a temp file and replacing after the save. The difficulty was to read data from the original file that was not in memory to place in the newly saved file.

Commit:
bug correction when reading region file with empty chunk

Commit:
completing a little bit the computing of light, but keeping a very simple approach, only taking care of vertical light propagation

Commit:
bug correction when reading a region file with empty chunks

Commit:
corrected a bug in NBT strings : length indicated was string length instead of byte length so it was corrupting the NBT output when some 2 bytes encoded character were in the string

Commit:
remove warning message about not implemented entity to avoid too much verbose
added some information in chunk loading fail

Commit:
changed treemaps to hashmaps as it seems like a better suiting choice and hoping to improve performance
made public all info on number of chunks / blocks / etc

Commit:
added implementation of mapItem file and definition of mapItem colors.
not fully tested and for the moment only manage one shade of base color

Commit:
corrected a bug that would prevent a player from being able to pick up a generated item
finalized map item generation.
disabled check for map item default tiling as minecraft works well with maps that are not aligned on the default tiling grid.
added the creation of subdirectories (region and data) in the world directory, if needed.

Commit:
cleaned some ugly implementation of sub data in items and item frames

Commit:
removed a map item color that was indicated in the minecraft wiki but not supported
bytes resulting from a color are now the variant that is the base color. managing variants is still TODO

Commit:
added BlockData for flower pot, it's deprecated but still readable by minecraft and a way to handle at least some flower pots as we don't deal with block entities yet

Commit:
specified block data for stone bricks

Commit:
filled all block data but the potion ones

Commit:
Set the last played date at current date by default, to avoid having a generated map being at the very end of the minecraft map list and displaying a date of 01/01/1970

Commit:
changed the center of a map item created from scratch to avoid having a pointer appear when map is displayed in an item frame

pdinklag commented:
Sorry for replying so late, I have just seen this.
I'm pretty surprised and amazed to see this, I cannot remember going anywhere public with my _library_, I only used it for some minor projects for our server.

I'll have a look at this next week, but let me already say this: thank you! :dancers:

Commit:
set the inlove field of breedable mobs as optional, as it is sometimes absent in minecraft generated files (found one case with a villager), and as the default value is perfectly fine

Commit:
adjusted health reading for mob to be compatible with minecraft version 1.9

Commit:
implementation of many mobs and different kinds of minecarts
implemented the structure for items (as owned by or contained in an entity)

Commit:
cleant comments of entity classes
added generic reading and writing of block entities
implemented a few block entities

Commit:
corrected a bug that ended in never saving block entities
made the mobspawner format so it can be understood by 1.9 and 1.8 versions of minecraft

Commit:
When reading chunks, a not yet supported block id would crash the library. now it's generating a warning instead

developpeur2000 commented:
Hey Patrick,

thanks for the notice,

I have finished working on my project last week, so I added a few more 
commits, especially around entity and block entity support, the library 
is now quite close to be able to read and write any element of a 
minecraft map.

I haven't had time to make any doc though, but it's quite straightforward.

Your lib was not complete but the code was clear and there was no 
specific functionality mixed in the map reading code, so it was a good 
starting point for my need.

my project was about generating minecraft maps from french geographic 
data to recreate the real world, if you want I'll let you know when the 
project becomes public (or it should be annonced on 
https://twitter.com/MinecraftALAC)

have a good day !

David Frémont
[email] david@developpeur2000.com
[tel] 07.81.98.88.85

Le 13/04/2016 à 17:27, Patrick Dinklage a écrit :

> Sorry for replying so late, I have just seen this.
> I'm pretty surprised and amazed to see this, I cannot remember going 
> anywhere public with my /library/, I only used it for some minor 
> projects for our server.
> 
> I'll have a look at this next week, but let me already say this: thank 
> you! :dancers:
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> https://github.com/pdinklag/Minecraft/pull/1#issuecomment-209509213

---

L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

emxjay commented:
@pdinklag Would be great if you can also provide a proper license for this library. Cheers!

developpeur2000 commented:
@pdinklag emxjay is working on the same project I used your library for.
The project is now live and generated maps can be obtained from this url : https://minecraft.ign.fr
